ddtrace/tracer/span.go

```diff
@@ -169,6 +170,37 @@ func (s *span) setSamplingPriority(priority int, sampler samplernames.SamplerNam
 	s.setSamplingPriorityLocked(priority, sampler, rate)
 }
 
+// setUser sets the span user ID tag as well as some optional user monitoring tags depending on the configuration.
+// The function assumes that the span it is called on is the trace's root span.
+func (s *span) setUser(id string, cfg UserMonitoringConfig) {
+	tags := map[string]string{}
+	keys := []string{ext.UserEmail, ext.UserName, ext.UserScope, ext.UserRole, ext.UserSessionID}
+	vals := []string{cfg.email, cfg.name, cfg.scope, cfg.role, cfg.sessionID}
+	// Build optional tags map before locking
+	for i, v := range vals {
+		if v != "" {
+			tags[keys[i]] = v
+		}
+	}
+	s.Lock()
+	defer s.Unlock()
+	if cfg.propagateID {
+		// Delete usr.id from the tags since _dd.p.usr.id takes precedence
+		delete(s.Meta, ext.UserID)
+		id = base64.StdEncoding.EncodeToString([]byte(id))
+		s.context.trace.setPropagatingTag(keyPropagatedUserID, id)
+	} else {
+		// Unset the propagated user ID so that a propagated user ID coming from upstream won't be propagated anymore.
+		delete(s.context.trace.propagatingTags, keyPropagatedUserID)
+		delete(s.Meta, keyPropagatedUserID)
+		// setMeta is used since the span is already locked
+		s.setMeta(ext.UserID, id)
```

Comment:
I think we've shot ourselves in the foot here with a potential bug in the API. By exposing `ext.UserID` in the public API, the user might (and definitely will) eventually set this tag directly and expect the behaviour to work. If the user does:
```go
span.SetTag(ext.UserID, id)
```
This should be expected to work correctly? Probably yes, as long as the API is there.

In the current form, it won't work as expected. For that reason, I suggest removing all the constants from the `ext` package because the user is not supposed to set any of them directly. Is that correct?

Reply:
Seems okay to me, the initial purpose of SetUser() is to make tags standardisation easier by setting the tags for the user so this is one more step in this direction. I can remove the tags from ext and define them in span.go (unexported) instead 👍

Reply:
Wonderful! Thanks! Ping me again for another review.